File: core/src/com/blakephillips/engine/ecs/systems/ai/JobSystem.java
package com.blakephillips.engine.ecs.systems.ai;

import com.badlogic.ashley.core.ComponentMapper;
import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.Family;
import com.badlogic.ashley.systems.IteratingSystem;
import com.badlogic.gdx.Gdx;
import com.blakephillips.engine.ai.State;
import com.blakephillips.engine.ecs.components.ai.JobComponent;
import com.blakephillips.engine.ecs.components.ai.StateComponent;

public class JobSystem extends IteratingSystem {

    ComponentMapper<JobComponent> jobComponents = ComponentMapper.getFor(JobComponent.class);
    public JobSystem() {
        super(Family.all(JobComponent.class).get());
    }
    @Override
    protected void processEntity(Entity entity, float v) {
        JobComponent job = jobComponents.get(entity);
        if (job.status == JobComponent.JobStatus.START_PENDING) {
            job.getStateEntity().add(new StateComponent(job.getCurrentState()));
            getEngine().addEntity(job.getStateEntity());
            job.status = JobComponent.JobStatus.RUNNING;
        }

        if (job.status == JobComponent.JobStatus.RUNNING) {
            if (job.getCurrentState().stateStatus == State.StateStatus.COMPLETE) {
                if (job.getCurrentState().getNextState() == null) {
                    job.getStateEntity().remove(StateComponent.class);
                    job.status = JobComponent.JobStatus.FINISHED;
                }

                if (job.getCurrentState().getNextState() != null) {
                    job.getStateEntity().add(new StateComponent(job.getCurrentState().getNextState()));
                    job.setCurrentState(job.getCurrentState().getNextState());
                }
            }
        }

        //just quit cancelled and incomplete jobs for now
        if (job.status == JobComponent.JobStatus.FINISHED
                || job.status == JobComponent.JobStatus.CANCELLED
                || job.status == JobComponent.JobStatus.INCOMPLETE) {
            getEngine().removeEntity(job.getStateEntity());
            getEngine().removeEntity(entity);
            Gdx.app.log("Game", String.format("Finished job %s", job.getName()));
        }
    }
}

File: core/src/com/blakephillips/engine/ecs/systems/ai/JobSystem.kt
package com.blakephillips.engine.ecs.systems.ai

import com.badlogic.ashley.core.ComponentMapper
import com.badlogic.ashley.core.Entity
import com.badlogic.ashley.core.Family
import com.badlogic.ashley.systems.IteratingSystem
import com.badlogic.gdx.Gdx
import com.blakephillips.engine.ai.State.StateStatus
import com.blakephillips.engine.ecs.components.ai.JobComponent
import com.blakephillips.engine.ecs.components.ai.StateComponent
import com.blakephillips.game.data.JobStatus

class JobSystem : IteratingSystem(Family.all(JobComponent::class.java).get()) {
    private var jobComponents = ComponentMapper.getFor(JobComponent::class.java)
    override fun processEntity(entity: Entity, v: Float) {
        val job = jobComponents[entity]

        when (job.status) {
            JobStatus.START_PENDING -> {
                job.stateEntity.add(StateComponent(job.currentState))
                engine.addEntity(job.stateEntity)
                job.status = JobStatus.RUNNING
            }

            JobStatus.RUNNING -> {
                when (job.currentState.stateStatus) {
                    StateStatus.COMPLETE -> {
                        if (job.currentState.nextState == null) {
                            job.stateEntity.remove(StateComponent::class.java)
                            job.status = JobStatus.FINISHED
                        }
                        if (job.currentState.nextState != null) {
                            job.stateEntity.add(StateComponent(job.currentState.nextState))
                            job.currentState = job.currentState.nextState
                        }
                    }

                    StateStatus.FAILED -> {
                        Gdx.app.debug("JobSystem", "Job '${job.name}' failed")
                        job.stateEntity.remove(StateComponent::class.java)
                        job.status = JobStatus.INCOMPLETE
                    }

                    else -> {}
                }
            }

            JobStatus.FINISHED, JobStatus.CANCELLED, JobStatus.INCOMPLETE -> {
                engine.removeEntity(job.stateEntity)
                engine.removeEntity(entity)
                Gdx.app.log("Game", "Finished job ${job.name}")
            }

            JobStatus.IDLE, null -> {}
        }
    }
}